### DISCUSSION TITLE
Git Pre Receive Hook Contiguous Response

### DISCUSSION THREAD

**Discussion by samuelsih** [dokku/dokku#7303] [Q&A]
Hello.

I tried to create my custom pre receive hook. I just want to know how you can make the remote output after `git push dokku master` contiguous.
I already see some of dokku's bash script and didn't find out how to do it.

Example:

```bash
#!/bin/bash
echo "Print this first"
sleep 2
echo "End"
```

I want to see string `Print this first` and after 2 seconds, string `End`. 
But in reality, `Print This First` and `End` came after 2 seconds. Not like dokku response where we can literally see the live output.

I have tried some solutions on the internet, but none of them works.

1. Use stdbuf
```bash
stdbuf -oL echo "Starting pre-receive hook..."
# Do something
stdbuf -oL echo "Pre-receive hook completed."
```

2. Redirect to some file descriptor
```bash
echo -n "Starting pre-receive hook..." >&2
# Do something
echo -n "Pre-receive hook completed." >&2
```

<br/>

Can you explain this to me? Any resource/keyword will be really helpful.

Thank you

**Comment by josegonzalez** [comment]
Converted this to a discussion.

What exactly are you trying to do? Are you trying to build something in Dokku, or something completely custom unrelated to Dokku?

**Reply by samuelsih** [reply]
Thank you for your response.

I made a project that has nothing to do with the dokku platform. 
I was just learning to implement the git protocol server with go language such as upload-pack, receive-pack, info-refs, and so on.  
Then it turns out, git has hooks features, for example pre-receive and post-receive.

Then I tried to setup server side pre receive hooks like this

```bash
echo "Start"
for i in {1..10}
do
  echo $i
  sleep 1
done
echo "End"
```

I tried the script above, and it worked as I expected. but I was not satisfied, because the remote response from the server came out almost at the same time. Like this example:

![PowerShell 2024-10-24 04-48-33(1)](https://github.com/user-attachments/assets/6788b45f-02f0-4ef7-929e-65ec0d4c3dc8)


I then remembered that once I had tried Dokku and it could do what I wanted, which was a contiguous response. Not the blocking response above. 
The Dokku's source code that i read, i can only see normal print `echo "Something"` with some extra `>&2` on some cases.

Sorry if there is any confusion in my explanation.
Thank you.

**Reply by josegonzalez** [reply]
Maybe this will help? https://stackoverflow.com/a/44955524

Might be a line ending issue with the file that is running the hook?

**Reply by samuelsih** [reply]
I already make sure that all my files are stored as LF format. Nothing works.

Just want to confirm, the pre receive hook script just use plain echo right? No tricks for contiguous response?
If that's right, then i will assume this problem is out of my control.

Thank you

**Reply by josegonzalez** [reply]
Yeah we don't do anything special here. [Here](https://github.com/dokku/dokku/blob/master/plugins/git/internal-functions#L355-L372) is how its done.

**Reply by samuelsih** [reply]
Thank you for the confirmation. I will close the discussion for now.

**Comment by mqnu00** [comment]
我发现和缓冲区有点关系，我在pre-receive-hook使用了 npm install --verbose 大量的输出挤爆了输出缓冲区，在客户端看上去这个命令的输出就一部分一部分传输过来了。
